Added middleware to supply Response with "x-correlation-id" header

lkurzyniec opened the pull request:
* added asp.net core middleware to supply `Response` with "x-correlation-id" header
* total fix for same `correlationId` in `CorrelationIdHeaderEnricher`
* added `CorrelationIdEnricherBase` base class
* added .editorconfig file

still to do:
* kind of middleware, kind of interceptor to rewrite `x-correlation-id` header in full framework

Commit:
Added middleware to supply Response with "x-correlation-id" header

Commit:
added base class CorrelationIdEnricherBase

lkurzyniec commented:
It should also address #34 issue.

Commit:
added .editorconfig file

lkurzyniec commented:
Please check `src/Serilog.Enrichers.CorrelationId/Enrichers/CorrelationIdEnricherBase.cs` file because I'm not sure that I implemented it properly, I haven't worked with different `TargetFrameworks` before (I mean both full framework and core framework).

alexjamesbrown commented on src/Serilog.Enrichers.CorrelationId/Middlewares/CorrelationIdHeaderSupplierMiddleware.cs:
```diff
@@ -0,0 +1,44 @@
+
+        public Task Invoke(HttpContext httpContext)
+        {
+            string correlationId = Guid.NewGuid().ToString();
```
This should be null until the below if() evaluates, as currently, it could be overwritten, making this allocation pointless

lkurzyniec replied:
fixed in d667c8e

alexjamesbrown commented on src/Serilog.Enrichers.CorrelationId/Middlewares/CorrelationIdHeaderSupplierMiddleware.cs:
```diff
@@ -0,0 +1,44 @@
+            string correlationId = Guid.NewGuid().ToString();
+            if (httpContext.Request.Headers.TryGetValue(_headerKey, out var values))
+            {
+                correlationId = values.First();
```
as per above comment, this could then be `correlationId = values.First() ?? Guid.NewGuid().ToString();`

lkurzyniec replied:
fixed in d667c8e

Commit:
avoid allocation

lkurzyniec commented:
@ejcoyle88 any update here?

ejcoyle88 commented:
Not yet sorry, I've been overloaded recently. I'll try to get to this this week.

sabbadino commented:
I suggest an option to add correlation id to to the request headers as well if not already present (similar to adding it to the response headers, as already implemented)

lkurzyniec commented:
@sabbadino adding correlation id to the request gives us nothing. you have to take care to pass it further. or you thinking about sth else

sabbadino commented:
Hi,
I'll take care to pass it further... I've got a wrapper around httpclient
for that.

Given the current implementation , to forward it , I should pick it up from
the response headers  which doesn't sound very "elegant'..
my 2 cents


On Sun, Dec 1, 2019, 22:01 Łukasz Kurzyniec <notifications@github.com>
wrote:

> @sabbadino <https://github.com/sabbadino> adding correlation id to the
> request gives us nothing. you have to take care to pass it further. or you
> thinking about sth else
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ekmsystems/serilog-enrichers-correlation-id/pull/36?email_source=notifications&email_token=ACHKZFF4X3XC6O7Y42KSGPTQWQQ3PA5CNFSM4JNTA3FKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFRVOTY#issuecomment-560158543>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACHKZFBHFWOTMXSAPZVYPBTQWQQ3PANCNFSM4JNTA3FA>
> .
>

lkurzyniec commented:
@sabbadino you are right, taking correlation id from response doesn't look well, +1 for you :)
I would like to also avoid modify request, IMHO it should be untouched.
How about adding correlation id to `ContextAccessor.HttpContext.Items`? would it satisfy you? ;)

sabbadino commented:
Sounds good to me :)

Commit:
added CorrelationIdConstants,
middleware add correlationid to HttpContext.Items,
added unit tests from middleware

lkurzyniec commented:
@ejcoyle88 any info here?

lkurzyniec commented:
:disappointed:
